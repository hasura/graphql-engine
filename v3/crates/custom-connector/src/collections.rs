use std::collections::BTreeMap;

use ndc_models;

use crate::{
    query::Result,
    state::{AppState, Row},
};

pub mod actors;
pub mod actors_by_movie;
pub mod countries;
pub mod institutions;
pub mod movies;
pub mod movies_by_actor_name;

pub(crate) fn get_collections() -> Vec<ndc_models::CollectionInfo> {
    vec![
        actors::collection_info(),
        movies::collection_info(),
        countries::collection_info(),
        institutions::collection_info(),
        actors_by_movie::collection_info(),
        movies_by_actor_name::collection_info(),
    ]
}

pub(crate) fn get_collection_by_name(
    collection_name: &ndc_models::CollectionName,
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    match collection_name.as_str() {
        "actors" => actors::rows(arguments, state),
        "countries" => countries::rows(arguments, state),
        "movies" => movies::rows(arguments, state),
        "institutions" => institutions::rows(arguments, state),
        "actors_by_movie" => actors_by_movie::rows(arguments, state),
        "movies_by_actor_name" => movies_by_actor_name::rows(arguments, state),
        _ => super::functions::get_function_by_name(collection_name, arguments, state),
    }
}
