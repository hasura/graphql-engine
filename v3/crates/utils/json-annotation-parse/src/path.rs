use crate::types::Value;
use jsonpath::{JSONPath, JSONPathElement};

#[derive(Debug)]
pub enum WalkError<Ann> {
    PathValueMismatch(JSONPathElement, Value<Ann>),
    MissingIndex(usize, Value<Ann>),
    MissingKey(String, Value<Ann>),
}

// reverse the path then walk it
pub fn walk<'a, Ann: Clone>(
    value: &'a Value<Ann>,
    path: &mut JSONPath,
) -> Result<&'a Ann, WalkError<Ann>> {
    let mut reversed_path = JSONPath(path.0.iter().rev().cloned().collect());
    walk_inner(value, &mut reversed_path)
}

fn walk_inner<'a, Ann: Clone>(
    value: &'a Value<Ann>,
    path: &mut JSONPath,
) -> Result<&'a Ann, WalkError<Ann>> {
    match path.0.pop() {
        Some(tail) => {
            match (value, &tail) {
                (Value::Array(_, items), JSONPathElement::Index(i)) => {
                    match items.get(*i) {
                        Some(item) => {
                            // got it!
                            walk_inner(item, path)
                        }
                        None => Err(WalkError::MissingIndex(*i, value.clone())),
                    }
                }
                (Value::Object(_, items), JSONPathElement::Key(s)) => {
                    match items.get(s) {
                        Some(item) => {
                            // got it!
                            walk_inner(item, path)
                        }
                        None => Err(WalkError::MissingKey(s.to_string(), value.clone())),
                    }
                }
                _ => Err(WalkError::PathValueMismatch(tail.clone(), value.clone())),
            }
        }
        None => {
            // we are here, return outer annotation
            match value {
                Value::Bool(ann, _) => Ok(ann),
                Value::Null(ann) => Ok(ann),
                Value::Number(ann, _) => Ok(ann),
                Value::String(ann, _) => Ok(ann),
                Value::Array(ann, _) => Ok(ann),
                Value::Object(ann, _) => Ok(ann),
            }
        }
    }
}
