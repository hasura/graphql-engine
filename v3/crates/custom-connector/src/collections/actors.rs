use std::collections::BTreeMap;

use axum::{http::StatusCode, Json};
use ndc_models;

use crate::{
    arguments::check_all_arguments_used,
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "actors".into(),
        description: Some("A collection of actors".into()),
        collection_type: "actor".into(),
        arguments: BTreeMap::new(),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "ActorByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    check_all_arguments_used(arguments)?;
    Ok(state.actors.values().cloned().collect())
}

pub(crate) fn filter_actors_by_name<'a>(
    state: &'a AppState,
    filter_first_name: Option<&'a str>,
    filter_last_name: Option<&'a str>,
) -> impl std::iter::Iterator<Item = Result<&'a Row>> {
    state
        .actors
        .values()
        .map(|actor| {
            let actor_name = actor
                .get("name")
                .ok_or_else(|| {
                    (
                        StatusCode::BAD_REQUEST,
                        Json(ndc_models::ErrorResponse {
                            message: "actor missing name".into(),
                            details: serde_json::Value::Null,
                        }),
                    )
                })?
                .as_str()
                .ok_or_else(|| {
                    (
                        StatusCode::BAD_REQUEST,
                        Json(ndc_models::ErrorResponse {
                            message: "actor name must be a string".into(),
                            details: serde_json::Value::Null,
                        }),
                    )
                })?;
            let names = actor_name.split(' ').collect::<Vec<_>>();
            let actor_first_name = names
                .first()
                .map_or(String::new(), std::string::ToString::to_string);
            let actor_last_name = names.into_iter().skip(1).collect::<Vec<_>>().join(" ");

            Ok((actor_first_name, actor_last_name, actor))
        })
        .filter_map(move |result| match result {
            Ok((actor_first_name, actor_last_name, actor)) => {
                if filter_first_name.map_or(true, |first_name| first_name == actor_first_name)
                    && filter_last_name.map_or(true, |last_name| last_name == actor_last_name)
                {
                    Some(Ok(actor))
                } else {
                    None
                }
            }
            Err(err) => Some(Err(err)),
        })
}
