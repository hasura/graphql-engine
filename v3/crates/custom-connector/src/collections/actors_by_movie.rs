use std::collections::BTreeMap;

use ndc_models;

use crate::{
    arguments::{check_all_arguments_used, parse_i32_argument},
    query::Result,
    state::{AppState, Row},
    types::actor::get_actor_movie_id,
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "actors_by_movie".into(),
        description: Some("Actors parameterized by movie".into()),
        collection_type: "actor".into(),
        arguments: BTreeMap::from_iter([(
            "movie_id".into(),
            ndc_models::ArgumentInfo {
                argument_type: ndc_models::Type::Named { name: "Int".into() },
                description: None,
            },
        )]),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::new(),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    let mut arguments = arguments
        .iter()
        .map(|(k, v)| (k.clone(), v))
        .collect::<BTreeMap<_, _>>();
    let movie_id_int = parse_i32_argument("movie_id", &mut arguments)?;
    check_all_arguments_used(&arguments)?;

    let mut actors_by_movie = vec![];

    for actor in state.actors.values() {
        let actor_movie_id_int = get_actor_movie_id(actor)?;
        if actor_movie_id_int == movie_id_int {
            actors_by_movie.push(actor.clone());
        }
    }

    Ok(actors_by_movie)
}
