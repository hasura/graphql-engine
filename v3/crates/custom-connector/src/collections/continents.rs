use std::collections::BTreeMap;

use ndc_models;

use crate::{
    arguments::check_all_arguments_used,
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "continents".into(),
        description: Some("A collection of continents".into()),
        collection_type: "continent".into(),
        arguments: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "ContinentByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    check_all_arguments_used(arguments)?;
    Ok(state.continents.values().cloned().collect())
}
