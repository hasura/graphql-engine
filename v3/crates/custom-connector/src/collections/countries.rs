use std::collections::BTreeMap;

use ndc_models;

use crate::{
    arguments::check_all_arguments_used,
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "countries".into(),
        description: Some("A collection of countries".into()),
        collection_type: "country".into(),
        arguments: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "CountryByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    check_all_arguments_used(arguments)?;
    Ok(state.countries.values().cloned().collect())
}
