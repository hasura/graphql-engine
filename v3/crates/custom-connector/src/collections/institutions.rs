use std::collections::BTreeMap;

use ndc_models;

use crate::{
    arguments::check_all_arguments_used,
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "institutions".into(),
        description: Some("A collection of institutions".into()),
        collection_type: "institution".into(),
        arguments: BTreeMap::new(),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "InstitutionByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    check_all_arguments_used(arguments)?;
    Ok(state.institutions.values().cloned().collect())
}

pub struct Institution<'a> {
    pub name: Option<&'a str>,
    pub location: Option<Location<'a>>,
    pub staff: Option<Vec<Staff<'a>>>,
    pub departments: Option<Vec<&'a str>>,
}

pub struct Location<'a> {
    pub city: Option<&'a str>,
    pub country: Option<&'a str>,
    pub campuses: Option<Vec<&'a str>>,
}

pub struct Staff<'a> {
    pub first_name: Option<&'a str>,
    pub last_name: Option<&'a str>,
}

pub(crate) fn filter_by_institution_object_input<'a>(
    state: &'a AppState,
    institution_query: Institution<'a>,
) -> impl std::iter::Iterator<Item = Result<&'a Row>> {
    state
        .institutions
        .values()
        .map(move |institution| {
            let institution_data = parse_institution(institution)?;
            let pass = check_institution_query(&institution_query, &institution_data);
            Ok((institution, pass))
        })
        .filter_map(move |result| match result {
            Ok((institution, pass)) => {
                if pass {
                    Some(Ok(institution))
                } else {
                    None
                }
            }
            Err(err) => Some(Err(err)),
        })
}

fn check_institution_query<'a>(
    institution_query: &'a Institution<'a>,
    institution_data: &'a Institution<'a>,
) -> bool {
    // check name
    if let Some(query_name) = institution_query.name {
        if let Some(data_name) = institution_data.name {
            if data_name == query_name {
                return true;
            }
        }
    };

    // check location
    if let Some(query_location) = &institution_query.location {
        if let Some(data_location) = &institution_data.location {
            // location -> city
            if let Some(query_city) = query_location.city {
                if let Some(data_city) = data_location.city {
                    if data_city == query_city {
                        return true;
                    }
                }
            }
            // location -> country
            if let Some(query_country) = query_location.country {
                if let Some(data_country) = data_location.country {
                    if data_country == query_country {
                        return true;
                    }
                }
            }
            // location -> campuses
            if let Some(query_campuses) = &query_location.campuses {
                if let Some(data_campuses) = &data_location.campuses {
                    for query_campus in query_campuses {
                        for data_campus in data_campuses {
                            if query_campus == data_campus {
                                return true;
                            }
                        }
                    }
                }
            }
        }
    };

    // TODO: implement filters for the other fields

    false
}

pub fn parse_institution(institution: &Row) -> Result<Institution<'_>> {
    let name = institution.get("name").and_then(|name| name.as_str());

    let staff = institution
        .get("staff")
        .and_then(|staff| parse_staff_members(staff));

    let location = institution.get("location").and_then(parse_location);

    let departments = institution.get("departments").and_then(parse_departments);

    Ok(Institution {
        name,
        location,
        staff,
        departments,
    })
}

pub fn parse_institution_query(
    institution: &serde_json::Map<String, serde_json::Value>,
) -> Result<Institution<'_>> {
    let name = institution.get("name").and_then(|name| name.as_str());

    let staff = institution
        .get("staff")
        .and_then(|staff| parse_staff_members(staff));

    let location = institution.get("location").and_then(parse_location);

    let departments = institution.get("departments").and_then(parse_departments);

    Ok(Institution {
        name,
        location,
        staff,
        departments,
    })
}

fn parse_location(location: &serde_json::Value) -> Option<Location> {
    if *location == serde_json::Value::Null {
        return None;
    }
    let city = location.get("city").and_then(|n| n.as_str());
    let country = location.get("country").and_then(|n| n.as_str());
    let campuses = location.get("campuses").and_then(|n| n.as_array());
    if city.is_none() && country.is_none() && campuses.is_none() {
        None
    } else {
        let campuses = campuses.map(|campuses| {
            campuses
                .iter()
                .filter_map(|val| val.as_str())
                .collect::<Vec<_>>()
        });
        Some(Location {
            city,
            country,
            campuses,
        })
    }
}

fn parse_staff_members(value: &serde_json::Value) -> Option<Vec<Staff>> {
    if *value == serde_json::Value::Null {
        None
    } else {
        value.as_array().map(|value| {
            value
                .iter()
                .filter_map(parse_staff_member)
                .collect::<Vec<_>>()
        })
    }
}

fn parse_departments(value: &serde_json::Value) -> Option<Vec<&str>> {
    if *value == serde_json::Value::Null {
        None
    } else {
        value.as_array().map(|value| {
            value
                .iter()
                .filter_map(|value| value.as_str())
                .collect::<Vec<_>>()
        })
    }
}

fn parse_staff_member(staff: &serde_json::Value) -> Option<Staff> {
    if *staff == serde_json::Value::Null {
        return None;
    }
    let first_name = staff.get("first_name").and_then(|n| n.as_str());
    let last_name = staff.get("last_name").and_then(|n| n.as_str());
    if first_name.is_none() && last_name.is_none() {
        None
    } else {
        Some(Staff {
            first_name,
            last_name,
        })
    }
}
