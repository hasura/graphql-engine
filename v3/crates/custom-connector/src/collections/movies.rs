use std::collections::BTreeMap;

use ndc_models;

use crate::{
    arguments::check_all_arguments_used,
    query::Result,
    state::{AppState, Row},
};

pub(crate) fn collection_info() -> ndc_models::CollectionInfo {
    ndc_models::CollectionInfo {
        name: "movies".into(),
        description: Some("A collection of movies".into()),
        collection_type: "movie".into(),
        arguments: BTreeMap::new(),
        foreign_keys: BTreeMap::new(),
        uniqueness_constraints: BTreeMap::from_iter([(
            "MovieByID".into(),
            ndc_models::UniquenessConstraint {
                unique_columns: vec!["id".into()],
            },
        )]),
    }
}

pub(crate) fn rows(
    arguments: &BTreeMap<ndc_models::ArgumentName, serde_json::Value>,
    state: &AppState,
) -> Result<Vec<Row>> {
    check_all_arguments_used(arguments)?;
    Ok(state.movies.values().cloned().collect())
}
