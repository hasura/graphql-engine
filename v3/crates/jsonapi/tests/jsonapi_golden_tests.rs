//! Tests that run JSONAPI to see if it works

use hasura_authn_core::{Identity, Role};
use reqwest::header::HeaderMap;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::str::FromStr;
use std::sync::Arc;

#[test]
fn test_get_succeeding_requests() {
    insta::glob!("passing/**/*.txt", |path| {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_all() // this enables time and IO
            .build()
            .unwrap();

        runtime.block_on(async {
            let TestRequest {
                query,
                model_name,
                jsonapi_state,
                resolved_metadata,
            } = test_setup(path);

            // always test in `default` subgraph for now
            let path = format!("/default/{model_name}");

            let http_context = execute::HttpContext {
                client: reqwest::Client::new(),
                ndc_response_size_limit: None,
            };

            let result = jsonapi::handler_internal(
                Arc::new(HeaderMap::default()),
                Arc::new(http_context.clone()),
                Arc::new(create_default_session()),
                &jsonapi_state,
                &resolved_metadata,
                axum::http::method::Method::GET,
                axum::http::uri::Uri::from_str(&path).unwrap(),
                query,
            )
            .await;

            match result {
                Ok(result) => insta::assert_debug_snapshot!("result", result),
                Err(e) => panic!("expected success, instead got {e}"),
            }
        });
    });
}

#[test]
fn test_get_failing_requests() {
    insta::glob!("failing/**/*.txt", |path| {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_all() // this enables time and IO
            .build()
            .unwrap();

        runtime.block_on(async {
            let TestRequest {
                query,
                model_name,
                jsonapi_state,
                resolved_metadata,
            } = test_setup(path);

            // always test in `default` subgraph for now
            let path = format!("/default/{model_name}");

            let http_context = execute::HttpContext {
                client: reqwest::Client::new(),
                ndc_response_size_limit: None,
            };

            let result = jsonapi::handler_internal(
                Arc::new(HeaderMap::default()),
                Arc::new(http_context.clone()),
                Arc::new(create_default_session()),
                &jsonapi_state,
                &resolved_metadata,
                axum::http::method::Method::GET,
                axum::http::uri::Uri::from_str(&path).unwrap(),
                query,
            )
            .await;

            match result {
                Ok(_) => panic!("expected failure"),
                Err(e) => insta::assert_debug_snapshot!("error", e),
            }
        });
    });
}

struct TestRequest {
    query: jsonapi_library::query::Query,
    model_name: String,
    jsonapi_state: jsonapi::State,
    resolved_metadata: metadata_resolve::Metadata,
}

fn test_setup(path: &Path) -> TestRequest {
    let directory = path.parent().unwrap();
    let model_name = path.file_stem().unwrap().to_str().unwrap();

    let query_params = std::fs::read_to_string(path).unwrap_or_else(|error| {
        panic!(
            "{}: Could not read file {path:?}: {error}",
            directory.display()
        )
    });

    let jsonapi_query = jsonapi_library::query::Query::from_params(&query_params);

    // Setup test context
    let root_test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");
    let metadata_path = root_test_dir.join("static").join("metadata.json");

    let metadata_string = std::fs::read_to_string(metadata_path.clone()).unwrap_or_else(|error| {
        panic!(
            "{}: Could not read file {path:?}: {error}",
            metadata_path.display()
        )
    });

    let metadata_json_value = serde_json::from_str(&metadata_string).unwrap();

    let input_metadata: open_dds::Metadata =
        open_dds::traits::OpenDd::deserialize(metadata_json_value, jsonpath::JSONPath::new())
            .unwrap();

    let configuration = get_metadata_resolve_configuration();

    let (resolved_metadata, _) = metadata_resolve::resolve(input_metadata, &configuration)
        .unwrap_or_else(|error| {
            panic!(
                "{}: Could not resolve metadata: {error}",
                directory.display()
            )
        });

    let jsonapi_state = jsonapi::State::new(&resolved_metadata);

    TestRequest {
        query: jsonapi_query,
        model_name: model_name.to_string(),
        jsonapi_state,
        resolved_metadata,
    }
}

// we will need to allow tests to define their own session options at some point
// the way that the GraphQL tests defined in `crates/engine/tests/execution.rs` do
fn create_default_session() -> hasura_authn_core::Session {
    //return an arbitrary identity with role emulation enabled
    let authorization = Identity::admin(Role::new("admin"));
    let role = Role::new("admin");
    let role_authorization = authorization.get_role_authorization(Some(&role)).unwrap();

    role_authorization.build_session(HashMap::new())
}

fn get_metadata_resolve_configuration() -> metadata_resolve::configuration::Configuration {
    let unstable_features = metadata_resolve::configuration::UnstableFeatures {
        enable_order_by_expressions: false,
        enable_ndc_v02_support: false,
        enable_subscriptions: true,
        enable_jsonapi: true,
        enable_aggregation_predicates: false,
    };

    metadata_resolve::configuration::Configuration { unstable_features }
}
