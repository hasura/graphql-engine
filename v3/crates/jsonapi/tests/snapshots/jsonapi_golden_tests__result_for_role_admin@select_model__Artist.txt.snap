---
source: crates/jsonapi/tests/jsonapi_golden_tests.rs
expression: result
input_file: crates/jsonapi/tests/passing/select_model/Artist.txt
---
DocumentData {
    data: Some(
        Multiple(
            [
                Resource {
                    _type: "default_Artist",
                    id: "1",
                    attributes: {
                        "ArtistId": Number(3),
                        "Name": String("Aerosmith"),
                    },
                    relationships: None,
                    links: None,
                    meta: None,
                },
                Resource {
                    _type: "default_Artist",
                    id: "2",
                    attributes: {
                        "ArtistId": Number(4),
                        "Name": String("Alanis Morissette"),
                    },
                    relationships: None,
                    links: None,
                    meta: None,
                },
                Resource {
                    _type: "default_Artist",
                    id: "3",
                    attributes: {
                        "ArtistId": Number(5),
                        "Name": String("Alice In Chains"),
                    },
                    relationships: None,
                    links: None,
                    meta: None,
                },
                Resource {
                    _type: "default_Artist",
                    id: "4",
                    attributes: {
                        "ArtistId": Number(6),
                        "Name": String("Ant√¥nio Carlos Jobim"),
                    },
                    relationships: None,
                    links: None,
                    meta: None,
                },
                Resource {
                    _type: "default_Artist",
                    id: "5",
                    attributes: {
                        "ArtistId": Number(7),
                        "Name": String("Apocalyptica"),
                    },
                    relationships: None,
                    links: None,
                    meta: None,
                },
                Resource {
                    _type: "default_Artist",
                    id: "6",
                    attributes: {
                        "ArtistId": Number(8),
                        "Name": String("Audioslave"),
                    },
                    relationships: None,
                    links: None,
                    meta: None,
                },
            ],
        ),
    ),
    included: None,
    links: None,
    meta: None,
    jsonapi: None,
}
